### PR TITLE
feat(cli): redact known secrets before trace and eval uploads

### PR DESCRIPTION
**Pull request by xeophon**
## Overview

Replaces #882 with a from-scratch, minimal implementation of the same goal: secrets do not leave the machine on upload. One ~75-line util in the CLI instead of a 973-line preflight module plus SDK changes.

## What it does

`prime traces upload`, `prime eval push`, and the auto-push after `prime eval run` replace every known secret value with `[REDACTED]` before the bytes go out — in the JSONL lines and `--context` values, in the eval data and `--name`, and in the auto-push's metadata, metrics, and task type. Known values are:
- the process environment's credentials: values under credential-like names (`KEY|TOKEN|SECRET|PASSWORD|CREDENTIAL|COOKIE|AUTHORIZATION|AUTH|SIGNATURE|SIG|PAT` as the name's last word, optionally numbered — compound names are head-final, so `HF_TOKEN`, `X-Api-Key`, `PGPASSWORD`, `API_KEY_2` are credentials while `TOKEN_URL`, `COOKIE_DOMAIN`, `KEY_FILE`, `SSH_AUTH_SOCK`, `KEYCLOAK_REALM` are not), plus, for any value that is a URL (`scheme://host…`, never prose), the password (or bare user token) of its userinfo and its credential-named query values (`DATABASE_URL`, an authenticated `HTTP_PROXY`, `wss://…?token=…`); a value that is a JSON object is a mapping too and gets the same rules inside it (`DOCKER_AUTH_CONFIG`, a service-account blob); 8+ chars
- the Prime API key
- `--secret <value>` or `--secret <file with one per line>`, repeatable, on `traces upload` and `eval push`

Matching is exact, over the serialized JSON text: each JSON string is decoded and then searched again for quoted JSON inside it, recursively, so any escape spelling (`\/`, uppercase hex, …) at any nesting depth is matched. Only JSON strings are touched, so numbers and structure cannot be corrupted, ordinary text is never rewritten, and a line with no hit is uploaded byte-identical, so content-addressed receipts still replay on rerun. Local files are never modified. The command reports how many occurrences were replaced (`.redacted` in JSON output).

## Deliberately not carried over from #882

- **Shape-based detection** (provider key regexes, header/cookie/assignment grammars, OpenAPI and JSON-schema awareness). Every review round grew the pattern set and its false-positive guards (`token_usage`, `oauth`, `BearerAuth`, …). That is a scanner's job: run trufflehog over the file before uploading if you want it. The CLI redacts what it knows, which is also [pi-share-hf](https://github.com/badlogic/pi-share-hf)'s position.
- **SDK-level auto-redaction, `push_evaluation`, exact batching and async upload rewrites, the fingerprint API** in prime-evals. The SDKs stay transports; redaction is a CLI concern, so no prime-evals release is needed and verifiers keeps no dependency on it.
- **Temp-file staging** for `traces upload`: `TracesClient.upload_lines` already exists, so lines are redacted in a generator on the way to the batcher.

## Companion

PrimeIntellect-ai/verifiers#2508 handles what only verifiers knows (client headers, harness env, resolved API keys, per-rollout tokens) with the same redactor. The two copies are intentionally independent so neither repo waits on the other's release.

## Tests

`test_redact.py` covers escaping at several depths (including `\/` and uppercase hex), the number/structure guarantee, and the secret sources. `test_traces_command.py` and `test_eval_push.py` check the commands end to end through Typer with stubbed clients: bytes untouched when nothing matches, secrets from env / API key / `--secret` literal / `--secret` file / `--context` / `--name` gone, local files unchanged.

---

> [!NOTE]
> **Medium Risk**
> Changes what leaves the machine on upload (security-sensitive); redaction is best-effort on known values only, so missed secrets or edge-case false negatives remain possible despite extensive tests.
> 
> **Overview**
> Adds a small **`redact`** helper and wires it into **`prime traces upload`**, **`prime eval push`**, and the post-run eval auto-push so outbound payloads are scrubbed before they hit the platform.
> 
> **Secret sources** are credential-like env vars (plus URL/JSON-nested credentials), the Prime API key, and repeatable **`--secret`** (literal or one-per-line file). Matches are **exact** inside JSON strings only (including nested quoted JSON), replaced with **`[REDACTED]`**; local files stay unchanged. Commands print a yellow notice and JSON output includes **`redacted`** counts where applicable.
> 
> **Traces** now stream through **`upload_lines`** with per-line redaction instead of uploading the raw file bytes unchanged. **Eval push** redacts loaded eval data and CLI fields such as **`run_id`** and **`name`** so secrets in identifiers cannot leak.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bf4475fcd8faad5d6fb2742f651727bfa4d3575e. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: packages/prime/src/prime_cli/commands/traces.py [hunk 1]
```diff
@@ -110,8 +122,9 @@ def on_batch(batch: Batch, receipt: UploadReceipt) -> None:
 
     try:
         client = _traces_client()
-        receipts = client.upload_file(
-            file,
+        redactor = Redactor(known_secrets(client.client.api_key, secret_args=secret))
+        receipts = client.upload_lines(
+            (redactor.json(line.decode()).encode() for line in read_jsonl_lines(file)),
             line_format=line_format,
             context=_parse_context(context),
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/commands/traces.py]: Fixed in ad008f3: `--context` values now go through the same `Redactor` (`redactor.value(_parse_context(context))`) before `upload_lines`; the command test asserts that `-c note=<secret>` arrives as `[REDACTED]`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    def __init__(self, secrets: Iterable[str]) -> None:
+        # Inside a JSON string a secret is escaped; inside a JSON document quoted within
+        # a string (a tool result) it is escaped twice. Match every spelling.
+        forms = set(secrets)
+        for _ in range(2):
+            forms |= {
+                json.dumps(form, ensure_ascii=escape)[1:-1]
+                for form in list(forms)
+                for escape in (True, False)
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in ccafa7e + ad008f3: the redactor now decodes each JSON string (`json.loads` on the token) before matching, so every valid escape spelling of the outer string is covered, and re-encodes only strings with a hit so untouched strings keep their exact bytes. For JSON documents quoted inside a string it generates the nested spellings (escaped once and twice, including the `\/` variant); the test covers a `\/`-escaped secret inside a quoted document.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/eval_push.py [hunk 3]
```diff
@@ -187,6 +189,12 @@ def push_eval_results_to_hub(
         }
         for sample in results_samples
     ]
+    eval_metadata, converted_results = redactor.value([eval_metadata, converted_results])
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/eval_push.py]: Fixed in ad008f3: `metrics` is redacted together with `eval_metadata` and `converted_results`, and `task_type` is read from the redacted `eval_metadata`, so every field sent by create, push, and finalize passes through the redactor.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: transfer-bulk-failures.jsonl [hunk 4]
```diff
@@ -0,0 +1 @@
+{"source": "a/app-a:v1", "platform": "linux/amd64"}
```
[the remark this answers is not in the document]
Comment (xeophon) [reply transfer-bulk-failures.jsonl]: Fixed in ad008f3: removed. It was a test artifact (`prime images transfer-bulk` writes its default `--failures-out` path into cwd) swept in by `git add -A`. The suite still recreates it at the repo root on every run, which is a separate hygiene issue.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+    def __init__(self, secrets: Iterable[str]) -> None:
+        # The string itself is decoded before matching. A JSON document quoted inside it
+        # (a tool result) is escaped once per nesting level, by an encoder that may also
+        # escape `/`: match those spellings too.
+        forms = set(secrets)
+        forms |= {form.replace("/", r"\/") for form in forms}
+        for _ in range(2):
+            forms |= {
+                json.dumps(form, ensure_ascii=escape)[1:-1]
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in bef6ef9: the redactor no longer enumerates escape spellings at all. Each JSON string is decoded (`json.loads`) and then searched again for quoted JSON inside it, recursively, so `ü`, `\/`, and any other valid escape are matched at any nesting depth; strings without a hit keep their exact bytes. `test_redact.py` covers uppercase hex and a doubly nested document.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/commands/evals.py [hunk 6]
```diff
@@ -1056,11 +1057,19 @@ def _push_single_eval(
     eval_id: Optional[str],
     is_public: bool = False,
     name: Optional[str] = None,
+    secrets: Iterable[str] = (),
 ) -> str:
     path = _validate_eval_path(config_path)
-    eval_data = _load_eval_directory(path)
+    api_client = APIClient()
+    redactor = Redactor(known_secrets(api_client.api_key, secret_args=secrets))
+    eval_data = redactor.value(_load_eval_directory(path))
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/commands/evals.py]: Fixed in bef6ef9: `--name` goes through the redactor together with the loaded eval data (`eval_data, name = redactor.value([...])`), so the name sent to `create_evaluation` is redacted; the test asserts `--name "run <secret>"` arrives as `run [REDACTED]`. `--env`, `--run-id`, and `--eval-id` stay as typed: they address existing resources and redacting them would break the request rather than protect anything.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/commands/evals.py [hunk 7]
```diff
@@ -1056,11 +1057,19 @@ def _push_single_eval(
     eval_id: Optional[str],
     is_public: bool = False,
     name: Optional[str] = None,
+    secrets: Iterable[str] = (),
 ) -> str:
     path = _validate_eval_path(config_path)
-    eval_data = _load_eval_directory(path)
+    api_client = APIClient()
+    redactor = Redactor(known_secrets(api_client.api_key, secret_args=secrets))
+    eval_data, name = redactor.value([_load_eval_directory(path), name])
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/commands/evals.py]: Fixed in 7751dc9, reversing my earlier reply: `env_slug`, `run_id`, and `eval_id` are now redacted together with the eval data and `--name`, up front, so the entire outbound request passes through the redactor with no carve-outs. A secret inside an identifier makes the request fail (resource not found) rather than leak it, which is what `--secret` promises. The test asserts `--run-id run-<secret>` arrives as `run-[REDACTED]`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/eval_push.py [hunk 8]
```diff
@@ -187,6 +189,14 @@ def push_eval_results_to_hub(
         }
         for sample in results_samples
     ]
+    eval_metadata, metrics, converted_results = redactor.value(
+        [eval_metadata, metrics, converted_results]
+    )
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/eval_push.py]: Fixed in 7751dc9: the auto-push redacts all of its inputs once, right after building the redactor — `env_name`, `model`, `job_id`, the resolved env slug/id, `metadata`, and the result rows — so everything downstream (the env lookup, every `create_evaluation` field, `push_samples`, `finalize_evaluation`) is built from redacted values.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        if not isinstance(value, str):
+            continue
+        if SECRET_NAME.search(name):
+            yield value
+        try:
+            parts = urlsplit(value)
+        except ValueError:
+            continue
+        if "@" in parts.netloc and (userinfo := parts.password or parts.username):
+            yield userinfo
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 5a5931a: the URL userinfo is registered both as written and percent-decoded (`unquote`), so `p%40ssword-0001` in the URL and `p@ssword-0001` as a client echoes it are both redacted. Test covers it; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+    would rewrite ordinary text; explicit values are taken as given."""
+    secrets = {
+        credential
+        for credential in env_credentials(os.environ)
+        if len(credential) >= MIN_SECRET_LENGTH
+    }
+    secrets.update(value for value in values if value)
+    for arg in secret_args:
+        path = Path(arg)
+        lines = path.read_text().splitlines() if path.is_file() else [arg]
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 5a5931a: the `--secret` file probe uses `os.path.isfile`, which returns False for anything that cannot be a path instead of raising, so a 400-character literal is taken as a literal. Test covers it.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+        if pattern is None:
+            return text
+
+        def string(match: re.Match[str]) -> str:
+            token = match.group(0)
+            try:
+                value = json.loads(token)
+            except ValueError:  # quotes in prose, not a JSON string
+                return token
+            redacted, hits = pattern.subn(REDACTED, self.json(value))
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:135]: Fixed in f940c45: `known_secrets` now refuses any value that is a substring of the `[REDACTED]` marker with a clear `ValueError`, so `--secret REDACTED` fails the upload instead of being kept inside the marker. Test covers it. Verifiers has no user-declared secrets, so there is nothing to refuse on that side.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+    secrets = {
+        credential
+        for credential in env_credentials(os.environ)
+        if len(credential) >= MIN_SECRET_LENGTH
+    }
+    secrets.update(value for value in values if value)
+    for arg in secret_args:
+        # `os.path.isfile` is False for anything that cannot be a path, a long literal included.
+        lines = Path(arg).read_text().splitlines() if os.path.isfile(arg) else [arg]
+        secrets.update(line.strip() for line in lines)
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 74e4d14: `--secret` values are no longer stripped — a literal is taken exactly as given, and a file entry loses only its line terminator (`splitlines` covers CRLF); empty lines are still dropped. The test asserts a literal with a trailing space and a CRLF file entry both register as written.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+import re
+from pathlib import Path
+from typing import Any, Iterable, Iterator, Mapping, Optional
+from urllib.parse import unquote, urlsplit
+
+REDACTED = "[REDACTED]"
+MIN_SECRET_LENGTH = 8
+SECRET_NAME = re.compile(
+    r"KEY|TOKEN|SECRET|PASSW|CREDENTIAL|COOKIE|AUTHORIZATION|(?:^|[_-])AUTH(?:[_-]|$)",
+    re.IGNORECASE,
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:24]: Fixed in 0800f8d: the credential words now have to be a whole segment of the name (delimited by `_`, `-`, or the ends), with `API_?KEY` kept as an explicit term for the unsegmented `APIKEY` spelling. `KEYCLOAK_REALM`, `COOKIECUTTER_CONFIG`, `TOKENIZERS_PARALLELISM`, and `GIT_AUTHOR_NAME` no longer match; `HF_TOKEN`, `X-Api-Key`, `OPENAI_APIKEY`, `Proxy-Authorization` still do. Test covers both sides; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    }
+    secrets.update(value for value in values if value)
+    for arg in secret_args:
+        # `os.path.isfile` is False for anything that cannot be a path, a long literal included.
+        # Taken as given: only a file entry's line terminator goes.
+        secrets.update(Path(arg).read_text().splitlines() if os.path.isfile(arg) else [arg])
+    secrets.discard("")
+    # No fixed marker can hide a value it contains; refuse rather than upload it.
+    if inside_marker := sorted(secret for secret in secrets if secret in REDACTED):
+        raise ValueError(f"cannot redact {inside_marker}: part of the {REDACTED} marker")
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 80b39fe: the marker check now applies only to explicit `--secret` values, which are refused; a discovered value inside the marker (`LEGACY_API_KEY=[REDACTED]`, `API_TOKEN=REDACTED`) is a sanitized placeholder and is skipped, so an ambient environment can never block an upload. Test covers both; verifiers#2508 skips such values the same way.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+        if pattern is None:
+            return text
+
+        def string(match: re.Match[str]) -> str:
+            token = match.group(0)
+            try:
+                value = json.loads(token)
+            except ValueError:  # quotes in prose, not a JSON string
+                return token
+            redacted, hits = pattern.subn(REDACTED, self.json(value))
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:135]: Not changed, deliberately. The normalization you describe happens only inside a quoted chunk that contained a secret and is therefore already being rewritten — it comes back with `[REDACTED]` in it, so nothing is silent — while every string and chunk without a hit keeps its exact bytes. Limiting recursion to chunks that parse as complete JSON documents would stop matching escaped secrets in JSON fragments embedded in prose (the case the recursion exists for), and re-dumping a whole nested document normalizes its escapes and separators too, so it does not buy fidelity. A redacted chunk with a normalized escape next to the marker is the smaller cost.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+        return
+    if "@" in parts.netloc:
+        # With a password slot, even an empty one, the username is a name, not a token.
+        userinfo = parts.username if parts.password is None else parts.password
+        if userinfo:
+            yield from {userinfo, unquote(userinfo)}
+    for pair in parts.query.split("&"):
+        name, _, raw = pair.partition("=")
+        if raw and SECRET_NAME.search(unquote(name)):
+            yield from {raw, unquote(raw)}
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 834dda9: credential-named query values register the `unquote_plus` spelling as well as the raw and `unquote` ones, so `?token=long+secret+value` also redacts `long secret value`. Test covers it; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+        if pattern is None:
+            return text
+
+        def string(match: re.Match[str]) -> str:
+            token = match.group(0)
+            try:
+                value = json.loads(token)
+            except ValueError:  # quotes in prose, not a JSON string
+                return token
+            redacted, hits = pattern.subn(REDACTED, self.json(value))
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:135]: Fixed in 834dda9, without a re-scan loop: the marker check now covers every way a secret can overlap the marker — inside it, or beginning with its tail (`]bar`) or ending with its head (`foo[`) — via `overlaps_marker`. Such a discovered value is skipped as a placeholder and such an explicit `--secret` is refused, so a replacement can never form a registered secret and a single pass stays sufficient. Test covers `]bar`, `foo[`, and `ED]tail`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+        name, _, raw = pair.partition("=")
+        if raw and SECRET_NAME.search(unquote(name)):
+            yield from {raw, unquote(raw), unquote_plus(raw)}
+
+
+def overlaps_marker(secret: str) -> bool:
+    """Whether replacing with the marker could leave or form `secret`: it sits inside
+    `[REDACTED]`, or begins with the marker's tail (`]bar`) or ends with its head
+    (`foo[`). Such a value is a placeholder or a pathological input, never a credential."""
+    return secret in REDACTED or any(
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in f1232cb: `overlaps_marker` now also rejects a secret that contains the whole marker, which completes the rule — the two overlap if one is inside the other or they share an edge — so a replacement can never form a registered secret and one pass stays sufficient. `foo[REDACTED]bar` is added to the refusal test; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+    explicit: set[str] = set()
+    for arg in secret_args:
+        # Taken as given: only a file entry's line terminator goes.
+        explicit.update(Path(arg).read_text().splitlines() if os.path.isfile(arg) else [arg])
+    explicit.discard("")
+    # No fixed marker can hide a value that overlaps it. A discovered one is a sanitized
+    # placeholder (`API_TOKEN=REDACTED`) and is skipped; a requested one is refused.
+    if colliding := sorted(secret for secret in explicit if overlaps_marker(secret)):
+        raise ValueError(f"cannot redact {colliding}: overlaps the {REDACTED} marker")
+    return {secret for secret in discovered if not overlaps_marker(secret)} | explicit
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 05cdc53 — this was my regression, thanks. The invariant is only that the marker itself contains no secret, so the check is back to `secret in REDACTED`: a discovered value inside the marker is skipped as a placeholder, an explicit one is refused, and a credential that contains or borders the marker (`live[REDACTED]pass-123`, `]bar`) is registered again. Leaving it out uploaded it for certain; registering it can miss only an instance that a replacement forms around another registered secret. Tested both ways.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:135]: Follow-up: 05cdc53 narrows this. Refusing every value that borders the marker turned out to drop real discovered credentials that contain `[REDACTED]` (see the later thread), so the rule is back to inside-the-marker only. The formed-instance case you described remains possible but needs another registered secret framed exactly by the halves of this one; that residual is accepted and documented rather than solved with a re-scan loop that would need its own termination bound.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+        if pattern is None:
+            return text
+
+        def string(match: re.Match[str]) -> str:
+            token = match.group(0)
+            try:
+                value = json.loads(token)
+            except ValueError:  # quotes in prose, not a JSON string
+                return token
+            redacted, hits = pattern.subn(REDACTED, self.json(value))
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:135]: Not changed, deliberately, and this is the settled position after the last two rounds pulled in opposite directions. Forming a registered secret across the marker boundary requires a registered secret that begins with `]` or ends with `[` — `]tail-secret-0002` in your example — which no minted or user-chosen credential does; only a contrived explicit `--secret` can produce it. Refusing such values (the previous round) dropped real credentials that contain `[REDACTED]`, so they are registered again and this residual is accepted and documented in `known_secrets`. A re-scan loop with a cycle-safe marker strategy is machinery for an input shape that does not occur, and adding it is exactly the growth this PR exists to reverse. If a maintainer wants the loop, it is a bounded three-line change in `Redactor.json`.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 21]
```diff
@@ -0,0 +1,131 @@
...
+import re
+from pathlib import Path
+from typing import Any, Iterable, Iterator, Mapping, Optional
+from urllib.parse import unquote, unquote_plus, urlsplit
+
+REDACTED = "[REDACTED]"
+MIN_SECRET_LENGTH = 8
+SECRET_NAME = re.compile(
+    r"(?:(?<![A-Za-z0-9])"
+    r"(?:API_?KEYS?|KEYS?|CREDENTIALS?|COOKIES?|AUTHORIZATION|AUTH|SIGNATURES?|SIG|PATS?)"
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py:21]: Fixed in 16ae2a0: `AUTH` now counts only as a name's last segment (`BASIC_AUTH`, `X-Auth`, bare `AUTH`), where it is what the variable holds; elsewhere it describes what the variable is about (`AUTH_TYPE`, `SSH_AUTH_SOCK`, `AUTH_URL`) and no longer matches. `Authorization` stays its own word. Test covers both sides; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+from pathlib import Path
+from typing import Any, Iterable, Iterator, Mapping, Optional
+from urllib.parse import unquote, unquote_plus, urlsplit
+
+REDACTED = "[REDACTED]"
+MIN_SECRET_LENGTH = 8
+SECRET_NAME = re.compile(
+    r"(?:(?<![A-Za-z0-9])"
+    r"(?:API_?KEYS?|KEYS?|CREDENTIALS?|COOKIES?|AUTHORIZATION|SIGNATURES?|SIG|PATS?)"
+    r"|TOKENS?|SECRETS?|PASSW(?:OR)?DS?)"
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 250a654 by generalising rather than excluding: compound names are head-final, so the credential word now has to be the name's last segment (optionally numbered, `API_KEY_2`), which is the rule already applied to `AUTH` extended to every word. `AUTHORIZATION_URL`, `COOKIE_DOMAIN`, `TLS_KEY_FILE`, `TOKEN_URL`, `PASSWORD_FILE` name metadata about a credential and no longer match; `HF_TOKEN`, `X-Api-Key`, `PGPASSWORD`, `BASIC_AUTH` still do, and URL values still yield their embedded credentials. The one accepted miss is a head word that is not a credential word (`SECRET_KEY_BASE`), stated in the docstring. `GOOGLE_APPLICATION_CREDENTIALS` still matches (its head word is CREDENTIALS); redacting that path from a trace is the low-harm side of the trade. Test covers both sides; mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: packages/prime/src/prime_cli/utils/redact.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+    for name, value in mapping.items():
+        if isinstance(value, Mapping):
+            yield from env_credentials(value)
+            continue
+        if not isinstance(value, str):
+            continue
+        if SECRET_NAME.search(name):
+            yield value
+        yield from url_credentials(value)
+        if value.startswith("{"):
```
[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Fixed in 0d40280: the guard is now `value.lstrip().startswith("{")`, so a JSON object value with leading whitespace is parsed like any other; the test's `DOCKER_AUTH_CONFIG` value now carries a leading space. Mirrored in PrimeIntellect-ai/verifiers#2508.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

[the remark this answers is not in the document]
Comment (xeophon) [reply packages/prime/src/prime_cli/utils/redact.py]: Correction: the test change landed in bf4475f (the fix itself was in the previous commit; the test edit had missed its anchor). The `DOCKER_AUTH_CONFIG` value in the test now starts with a space.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
